File: angular/src/app/core/services/task.service.ts
import { inject, Injectable } from '@angular/core';
import {
  Firestore,
  collection,
  collectionData,
  addDoc,
  deleteDoc,
  doc,
  updateDoc,
  query,
  where,
  serverTimestamp,
} from '@angular/fire/firestore';
import { Observable } from 'rxjs';
import { Task } from '../models/task.model';

@Injectable({ providedIn: 'root' })
export class TaskService {
  private firestore = inject(Firestore);

  /**
   * Tâches assignées à un élève (assigneeId === userId)
   */
  getTasksEleve$(eleveId: string): Observable<Task[]> {
    const q = query(
      collection(this.firestore, 'tasks'),
      where('assigneeId', '==', eleveId)
    );
    return collectionData(q, { idField: 'id' }) as Observable<Task[]>;
  }

  /**
   * Tâches créées par un bénévole (createurId === userId)
   */
  getTasksBenevole$(benevoleId: string): Observable<Task[]> {
    const q = query(
      collection(this.firestore, 'tasks'),
      where('createurId', '==', benevoleId)
    );
    return collectionData(q, { idField: 'id' }) as Observable<Task[]>;
  }

  createTask(task: Omit<Task, 'id'>): Promise<void> {
    return addDoc(collection(this.firestore, 'tasks'), {
      ...task,
      createdAt: serverTimestamp(),
    }).then(() => undefined);
  }

  deleteTask(taskId: string): Promise<void> {
    return deleteDoc(doc(this.firestore, `tasks/${taskId}`));
  }

  toggleTask(taskId: string, terminee: boolean): Promise<void> {
    return updateDoc(doc(this.firestore, `tasks/${taskId}`), { terminee });
  }
}

File: flutter/lib/core/services/task_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/task.dart';

/// Service de gestion des tâches — miroir de TaskService Angular.
class TaskService {
  TaskService({FirebaseFirestore? firestore})
      : _db = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _db;

  // ─── Lecture temps réel ───────────────────────────────────────────────────

  /// Tâches assignées à un élève (assigneeId == eleveId)
  Stream<List<Task>> getTasksEleve(String eleveId) {
    return _db
        .collection('tasks')
        .where('assigneeId', isEqualTo: eleveId)
        .snapshots()
        .map((snap) => snap.docs
            .map((d) =>
                Task.fromDoc(d as DocumentSnapshot<Map<String, dynamic>>))
            .toList());
  }

  /// Tâches créées par un bénévole (createurId == benevoleId)
  Stream<List<Task>> getTasksBenevole(String benevoleId) {
    return _db
        .collection('tasks')
        .where('createurId', isEqualTo: benevoleId)
        .snapshots()
        .map((snap) => snap.docs
            .map((d) =>
                Task.fromDoc(d as DocumentSnapshot<Map<String, dynamic>>))
            .toList());
  }

  // ─── Écriture ─────────────────────────────────────────────────────────────

  Future<void> createTask(Task task) async {
    await _db.collection('tasks').add({
      ...task.toMap(),
      'createdAt': FieldValue.serverTimestamp(),
    });
  }

  Future<void> deleteTask(String taskId) async {
    await _db.collection('tasks').doc(taskId).delete();
  }

  Future<void> toggleTask(String taskId, {required bool terminee}) async {
    await _db.collection('tasks').doc(taskId).update({'terminee': terminee});
  }
}
